src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/JsonMetadataServices.Converters.cs

```diff
@@ -264,13 +264,39 @@ public static JsonConverter<T> GetUnsupportedTypeConverter<T>()
                 ThrowHelper.ThrowArgumentNullException(nameof(underlyingTypeInfo));
             }
 
-            JsonConverter<T>? underlyingConverter = underlyingTypeInfo.Converter as JsonConverter<T>;
-            if (underlyingConverter == null)
+            JsonConverter<T> underlyingConverter = GetTypedConverter<T>(underlyingTypeInfo.Converter);
+
+            return new NullableConverter<T>(underlyingConverter);
+        }
+
+        /// <summary>
+        /// Creates a <see cref="JsonConverter{T}"/> instance that converts <typeparamref name="T?"/> values.
+        /// </summary>
+        /// <typeparam name="T">The generic definition for the underlying nullable type.</typeparam>
+        /// <param name="options">The <see cref="JsonSerializerOptions"/> to use for serialization and deserialization.</param>
+        /// <returns>A <see cref="JsonConverter{T}"/> instance that converts <typeparamref name="T?"/> values</returns>
+        /// <remarks>This API is for use by the output of the System.Text.Json source generator and should not be called directly.</remarks>
+        public static JsonConverter<T?> GetNullableConverter<T>(JsonSerializerOptions options) where T : struct
+        {
+            if (options is null)
             {
-                throw new InvalidOperationException(SR.Format(SR.SerializationConverterNotCompatible, underlyingConverter, typeof(T)));
+                ThrowHelper.ThrowArgumentNullException(nameof(options));
             }
 
+            JsonConverter<T> underlyingConverter = GetTypedConverter<T>(options.GetConverterFromTypeInfo(typeof(T)));
```

Comment:
it won't. The only place we do rooting is public GetConverter. This will lead to exception unless you called one of the rooting APIs. This is the reason I refactored `GetConverter*` methods so that we preserve this behavior

Reply:
I'm not disputing that, it's more a question of why this specific type needing to be different. It seems weird that it does and introduces more accidental complexity. Isn't it possible to emulate prior art?

Reply:
this is called from generated code and technically speaking we could call `options.GetConverter` but at the same time that currently roots reflection converters - as long as we do that that would break linker safety